### ISSUE TITLE
generate_data() is called every 10th iteration and - single entry validation set

### ISSUE THREAD

**Issue by alisiahkoohi** [adler-j/learned_primal_dual#2]
Why do you generate new training data pairs just every 10th iteration? For instance [here](https://github.com/adler-j/learned_primal_dual/blob/master/ellipses/learned_primal_dual.py#L162). 

Also [this line](https://github.com/adler-j/learned_primal_dual/blob/master/ellipses/learned_primal_dual.py#L171) suggests that the validation error is only being evaluated over a single data pair. So technically you validation set contains a single data pair.

**Comment by adler-j** [comment]
Both observations are true.

I did try generating data in every iteration, but it made no/very little difference at some performance cost. Now that you mention it I'm not sure if this made it into the article. Regardless, you can try generating at each iteration and see what happens.

Regarding the second point, the numerical values were indeed only evaluated for the image displayed in the article (the shepp-logan phantom). Notably, this is not a "random sample" from the prior but a rather special case. I picked it to show that the method can generalize quite well. This should be covered in the article.

**Comment by alisiahkoohi** [comment]
I ran `learned_primal_dual.py` for ellipses with `generate_data()` being called every iteration. I double checked to make sure I haven't changed anything else but the training loss blew up after ~5k steps.
Will try the original code.

![screenshot from 2018-07-18 11-33-02](https://user-images.githubusercontent.com/22459962/42891866-5df2f58a-8a7e-11e8-95a0-cb85af888466.png)

**Comment by adler-j** [comment]
Very interesting. Long time since I ran this code, does it work well if you call it every 10:th iteration?

**Comment by adler-j** [comment]
Did you ever come through to running the original code? Did it work?

**Comment by alisiahkoohi** [comment]
Sorry for the late response. I should mention that I was trying to run the code on CPU so maybe that is why I see this issue. If that is the case it may also explain why the original code didn't work for me.

I haven't got a chance to install ODL extensions to run it on GPU yet.

**Comment by ChengV0** [comment]
> 你有没有通过运行原始代码？它有用吗？

Running source code does not get the results in the article，What is the cause? I hope you can give me some advice.

**Comment by ChengV0** [comment]
> Did you ever come through to running the original code? Did it work?


![image](https://user-images.githubusercontent.com/34834140/49136359-5974d000-f324-11e8-8da9-79b8555d0032.png)

I'm sorry to have some problems uploading pictures. This is my result. I look forward to your reply. Thank you.

**Comment by adler-j** [comment]
Which file are you running to get those results? I've tried the current master on my machine and I get reasonable results. I also need to know e.g. how many iterations you ran.

Also, what version of ODL are you using?

**Comment by ChengV0** [comment]
> Which file are you running to get those results? I've tried the current master on my machine and I get reasonable results. I also need to know e.g. how many iterations you ran.

Thank you for your reply.I running the learned_primal_dual.py and.learned_chambolle_pock.py   The screenshot is the result of running the latter, and I stopped training because it collapsed at about 3000 steps. I'm sorry I didn't save the screenshots of loss and psnr.

**Comment by adler-j** [comment]
Sadly a tensorflow bug that hasn't been fixed in half a year (https://github.com/tensorflow/tensorflow/issues/16864) is causing this code to run extremely slowly on my machine, so it's hard for me to debug.

Are you running on master?

What version of ODL are you using?

**Comment by ChengV0** [comment]
> Which file are you running to get those results? I've tried the current master on my machine and I get reasonable results. I also need to know e.g. how many iterations you ran.
> 
> Also, what version of ODL are you using?



> > Which file are you running to get those results? I've tried the current master on my machine and I get reasonable results. I also need to know e.g. how many iterations you ran.
> 
> Thank you for your reply.I running the learned_primal_dual.py and.learned_chambolle_pock.py The screenshot is the result of running the latter, and I stopped training because it collapsed at about 3000 steps. I'm sorry I didn't save the screenshots of loss and psnr.

![image](https://user-images.githubusercontent.com/34834140/49140152-8201c780-f32e-11e8-9658-a2df640c8013.png)
This is the result of my operation about learned_primal_dual.py. Seen from the chart, the result is not up to expectation. Is it due to inadequate training?

**Comment by ChengV0** [comment]
> Sadly a tensorflow bug that hasn't been fixed in half a year ([tensorflow/tensorflow#16864](https://github.com/tensorflow/tensorflow/issues/16864)) is causing this code to run extremely slowly on my machine, so it's hard for me to debug.
> 
> Are you running on master?
> 
> What version of ODL are you using?



yes ,odl-1.0.0.dev0

**Comment by adler-j** [comment]
Are you getting the same problem with learned_primal_dual.py? E.g. can you show a loss curve

**Comment by ChengV0** [comment]
> Are you getting the same problem with learned_primal_dual.py? E.g. can you show a loss curve

![image](https://user-images.githubusercontent.com/34834140/49141322-4fa59980-f331-11e8-9bd7-2c5f00c6fb0e.png)
I think so.There is no tendency for the results to improve.

**Comment by adler-j** [comment]
What implementation of the raytransform are you using? 

E.g.

    >>> operator.impl
    'astra_cuda'

**Comment by ChengV0** [comment]
![image](https://user-images.githubusercontent.com/34834140/49141764-55e84580-f332-11e8-9abb-33e21b1ca1ad.png)
I just use the original code.

**Comment by adler-j** [comment]
I would _very_ strongly recommend you install astra, try

```
conda install -c astra-toolbox astra-toolbox
```

**Comment by adler-j** [comment]
My learning curves look like this:

![image](https://user-images.githubusercontent.com/2202312/49142114-5fed5280-f2f8-11e8-8116-739e26490d41.png)

E.g. seems to be improving just fine.

The above is for "learned_chambolle_pock.py"

**Comment by ChengV0** [comment]
> I would _very_ strongly recommend you install astra, try
> 
> ```
> conda install -c astra-toolbox astra-toolbox
> ```

![image](https://user-images.githubusercontent.com/34834140/49142260-533a2000-f333-11e8-8bd1-86dc235c4cbb.png)
Our results are different.

**Comment by ChengV0** [comment]
> My learning curves look like this:
> 
> ![image](https://user-images.githubusercontent.com/2202312/49142114-5fed5280-f2f8-11e8-8116-739e26490d41.png)
> 
> E.g. seems to be improving just fine.
> 
> The above is for "learned_chambolle_pock.py"

Maybe you can send me the code you're running now. Let me have a look. Linux encountered problems in installation astra-toolbox

**Comment by adler-j** [comment]
What problem did you encounter? I'm literally just running the code in this repo.

**Comment by ChengV0** [comment]
> What problem did you encounter? I'm literally just running the code in this repo.

It frustrates me that I can't reproduce your results using source code.

**Comment by ChengV0** [comment]
> My learning curves look like this:
> 
> ![image](https://user-images.githubusercontent.com/2202312/49142114-5fed5280-f2f8-11e8-8116-739e26490d41.png)
> 
> E.g. seems to be improving just fine.
> 
> The above is for "learned_chambolle_pock.py"

Why are there two curves on each graph, but mine has only one?

**Comment by adler-j** [comment]
training and testing losses (blue is train, orange is test)

**Comment by ChengV0** [comment]
> training and testing losses (blue is train, orange is test)

Oh， So the code you use is different from mine？

**Comment by adler-j** [comment]
The "learned_chambolle_pock.py" file includes both training and testing losses. See e.g.

https://github.com/adler-j/learned_primal_dual/blob/64901e8b585b23fcadae1b007e97fcc06b091f5e/ellipses/learned_chambolle_pock.py#L152-L154

**Comment by ChengV0** [comment]
> training and testing losses (blue is train, orange is test)

![image](https://user-images.githubusercontent.com/34834140/49143994-3acc0480-f337-11e8-9884-702baa3f5a7c.png)
![image](https://user-images.githubusercontent.com/34834140/49144021-49b2b700-f337-11e8-8c71-2aaa2b6967c2.png)
I got terrible results and I don't know why.（primal_dual.py）

**Comment by ChengV0** [comment]
> The "learned_chambolle_pock" file includes both training and testing losses. See e.g.
> 
> [learned_primal_dual/ellipses/learned_chambolle_pock.py](https://github.com/adler-j/learned_primal_dual/blob/64901e8b585b23fcadae1b007e97fcc06b091f5e/ellipses/learned_chambolle_pock.py#L152-L154)
> 
> Lines 152 to 154 in [64901e8](/adler-j/learned_primal_dual/commit/64901e8b585b23fcadae1b007e97fcc06b091f5e)
> 
>  test_summary_writer = tf.summary.FileWriter(adler.tensorflow.util.default_tensorboard_dir(name) + '/test', 
>                                              sess.graph) 
>  train_summary_writer = tf.summary.FileWriter(adler.tensorflow.util.default_tensorboard_dir(name) + '/train')

Well， thank you

**Comment by adler-j** [comment]
It's very hard for me to debug remotely, but my best guess right now is that you need astra.

Except for that, make sure that you have downloaded the laster version of this repo and ODL (e.g. re-install them).

Finally, what TF version do you run?

**Comment by ChengV0** [comment]
Ok, thank you very much for your patient answer. I will try again. TF:(1.8.0）

**Comment by adler-j** [comment]
It's great to get feedback. I try to make sure the code is runnable by everyone. please report any progress.

**Comment by ChengV0** [comment]
> It's great to get feedback. I try to make sure the code is runnable by everyone. please report any progress.

![image](https://user-images.githubusercontent.com/34834140/49192913-1feda480-f3b8-11e8-9071-91f243bd5ff2.png)
![image](https://user-images.githubusercontent.com/34834140/49192930-37c52880-f3b8-11e8-847c-2dd606792dd1.png)
The result is so bad(primal_dual.py)If it's convenient for you, I'd like to see your training summaries/images.Thank you.

**Comment by adler-j** [comment]
Did you re-install this library and ODL as advised? Did you install ASTRA? If you do not follow my advice it's hard to help.

My GPU is currently quite busy, but my training curve is a rather smooth convergence towards ~37 PSNR, nothing like what you are seeing.

**Comment by ChengV0** [comment]
> Did you re-install this library and ODL as advised? Did you install ASTRA? If you do not follow my advice it's hard to help.
> 
> My GPU is currently quite busy, but my training curve is a rather smooth convergence towards ~37 PSNR, nothing like what you are seeing.

![image](https://user-images.githubusercontent.com/34834140/49206483-2f3b1500-f3ed-11e8-9ee3-4044f84ba4fa.png)
Thank you for your reply.When I installed astra_toolbox, I met  difficulties.Because the server did not install conda but used pip3. 
May I have a look at your training pictures?（summaries/images）
In training, is your GPU-Util similar to mine?What is the reason for this?Thank you again.

**Comment by adler-j** [comment]
My curves look like this:

![image](https://user-images.githubusercontent.com/2202312/49207225-9a1c2a00-f3b4-11e8-96ed-679e35b36b9b.png)

![image](https://user-images.githubusercontent.com/2202312/49207095-4b6e9000-f3b4-11e8-85a1-2bd7935a07bd.png)

![image](https://user-images.githubusercontent.com/2202312/49207125-5d503300-f3b4-11e8-9957-b87b2374935f.png)

My GPU is way more busy than that, e.g. not 100% but far higher than 1%. I guess mine is busy because i have installed ASTRA, without it all the time is spent in scikit-image.

**Comment by ChengV0** [comment]
> My curves look like this:
> 
> ![image](https://user-images.githubusercontent.com/2202312/49207225-9a1c2a00-f3b4-11e8-96ed-679e35b36b9b.png)
> 
> ![image](https://user-images.githubusercontent.com/2202312/49207095-4b6e9000-f3b4-11e8-85a1-2bd7935a07bd.png)
> 
> ![image](https://user-images.githubusercontent.com/2202312/49207125-5d503300-f3b4-11e8-9957-b87b2374935f.png)
> 
> My GPU is way more busy than that, e.g. not 100% but far higher than 1%. I guess mine is busy because i have installed ASTRA, without it all the time is spent in scikit-image.

Thank you very much. You have helped me a lot.

**Comment by AceCoooool** [comment]
> I would _very_ strongly recommend you install astra, try
> 
> ```
> conda install -c astra-toolbox astra-toolbox
> ```

I also meet similar curve as @ChengV0  meets when run `ellipses/learned_primal.py` and `ellipses/learned_primal_dual.py` 

I think astra-toolbox is faster than skimage, however, it will not influence the results too much. For example: （addition: oh no!!! it will infulence the results!!!!! However, I did not know why? --- Using astra can get the results as author @adler-j 's learning curve. And using skimage get learning curve as ChengV0 meets. ）
```
import astra
import numpy as np
from skimage import measure
import scipy.io

P = scipy.io.loadmat('phantom.mat')['phantom256']

# astra
vol_geom = astra.create_vol_geom(256, 256)
proj_geom = astra.create_proj_geom('parallel', 1.0, 384, np.linspace(0, np.pi, 180, False))

proj_id = astra.create_projector('cuda', proj_geom, vol_geom)
sinogram_id, sinogram = astra.create_sino(P, proj_id)
rec_id = astra.data2d.create('-vol', vol_geom)
cfg = astra.astra_dict('FBP_CUDA')
cfg['ReconstructionDataId'] = rec_id
cfg['ProjectionDataId'] = sinogram_id
cfg['option'] = {'FilterType': 'Ram-Lak'}
alg_id = astra.algorithm.create(cfg)
astra.algorithm.run(alg_id)
rec = astra.data2d.get(rec_id)

print("psnr: ", measure.compare_psnr(P.astype(np.float32), rec))

astra.algorithm.delete(alg_id)
astra.data2d.delete(rec_id)
astra.data2d.delete(sinogram_id)
astra.projector.delete(proj_id)

# skimage
from skimage.transform import radon, iradon

img = P.astype(np.float64)
theta = np.linspace(0., 180., 180, endpoint=False)
sinogram = radon(img, theta=theta, circle=True)
reconstruction_fbp = iradon(sinogram, theta=theta, circle=True)

print("psnr: ", measure.compare_psnr(img, reconstruction_fbp))
```
psnr of astra: 34.125
psnr of skimage: 34.116

**Comment by AceCoooool** [comment]
@adler-j I have some questions（not open new issue for convience）:
1. the default geometry (`odl.tomo.parallel_beam_geometry`) using proj_space (geometry.det_partition.cell_sides) not equal to one. Is this better ? Due to many implementation (e.g. sklearn radon function) using 1.
2. The projection's derivation is back projection. However, many library (e.g. astra-toolbox and skimage radon) implement these not in full accord . 

Thank you @adler-j  !
